File: Dart/14-Backtracking/Algorithms/2-SudokuSolver/SudokuSolver.dart
import 'dart:math';

void main() {
  // Example Sudoku grid (0 represents empty cells)
  List<List<int>> grid = [
    [5, 3, 0, 0, 7, 0, 0, 0, 0],
    [6, 0, 0, 1, 9, 5, 0, 0, 0],
    [0, 9, 8, 0, 0, 0, 0, 6, 0],
    [8, 0, 0, 0, 6, 0, 0, 0, 3],
    [4, 0, 0, 8, 0, 3, 0, 0, 1],
    [7, 0, 0, 0, 2, 0, 0, 0, 6],
    [0, 6, 0, 0, 0, 0, 2, 8, 0],
    [0, 0, 0, 4, 1, 9, 0, 0, 5],
    [0, 0, 0, 0, 8, 0, 0, 7, 9],
  ];

  print("Original Sudoku Grid:");
  printGrid(grid);

  // 1. Backtracking with Simple Iteration
  List<List<int>> grid1 = List.generate(9, (i) => List.from(grid[i])); // Create a copy
  if (solveSudoku1(grid1)) {
    print("\nSolved Sudoku (Backtracking 1):");
    printGrid(grid1);
  } else {
    print("\nNo solution exists (Backtracking 1).");
  }

  // 2. Backtracking with Optimized Search
  List<List<int>> grid2 = List.generate(9, (i) => List.from(grid[i])); // Create a copy
  if (solveSudoku2(grid2)) {
    print("\nSolved Sudoku (Backtracking 2):");
    printGrid(grid2);
  } else {
    print("\nNo solution exists (Backtracking 2).");
  }

  // 3. Backtracking with Forward Checking
  List<List<int>> grid3 = List.generate(9, (i) => List.from(grid[i]));
  if (solveSudoku3(grid3)) {
    print("\nSolved Sudoku (Backtracking 3):");
    printGrid(grid3);
  } else {
    print("\nNo solution exists (Backtracking 3).");
  }

  // 4. Backtracking with MRV Heuristic
  List<List<int>> grid4 = List.generate(9, (i) => List.from(grid[i]));
  if (solveSudoku4(grid4)) {
    print("\nSolved Sudoku (Backtracking 4 - MRV):");
    printGrid(grid4);
  } else {
    print("\nNo solution exists (Backtracking 4 - MRV).");
  }

  // 5. Backtracking with Random Cell Selection
  List<List<int>> grid5 = List.generate(9, (i) => List.from(grid[i]));
  if (solveSudoku5(grid5)) {
    print("\nSolved Sudoku (Backtracking 5 - Random):");
    printGrid(grid5);
  } else {
    print("\nNo solution exists (Backtracking 5).");
  }
}

// Helper function to print the Sudoku grid
void printGrid(List<List<int>> grid) {
  for (int i = 0; i < 9; i++) {
    if (i % 3 == 0 && i != 0) {
      print("---------");
    }
    for (int j = 0; j < 9; j++) {
      if (j % 3 == 0 && j != 0) {
        print("|"); // Removed end: ""
      }
      print("${grid[i][j]} "); // Removed end: ""
    }
    print("");
  }
}

// 1. Backtracking with Simple Iteration
// This is the basic backtracking approach.  It iterates through each cell
// and tries numbers 1-9.  If a number is valid, it recurses.  If no number
// is valid, it backtracks.
bool solveSudoku1(List<List<int>> grid) {
  for (int row = 0; row < 9; row++) {
    for (int col = 0; col < 9; col++) {
      if (grid[row][col] == 0) {
        for (int number = 1; number <= 9; number++) {
          if (isValid(grid, row, col, number)) {
            grid[row][col] = number;
            if (solveSudoku1(grid)) {
              return true;
            } else {
              grid[row][col] = 0; // Backtrack
            }
          }
        }
        return false;
      }
    }
  }
  return true; // Grid is full, solution found
}

// 2. Backtracking with Optimized Search
// This version optimizes the search by finding the next empty cell
// directly, instead of iterating through the entire grid.  This can
// be slightly more efficient.
bool solveSudoku2(List<List<int>> grid) {
  int row = -1;
  int col = -1;
  bool isEmpty = true;
  for (int i = 0; i < 9; i++) {
    for (int j = 0; j < 9; j++) {
      if (grid[i][j] == 0) {
        row = i;
        col = j;
        isEmpty = false;
        break;
      }
    }
    if (!isEmpty) {
      break;
    }
  }

  if (isEmpty) {
    return true; // No empty cells, we are done
  }

  for (int number = 1; number <= 9; number++) {
    if (isValid(grid, row, col, number)) {
      grid[row][col] = number;
      if (solveSudoku2(grid)) {
        return true;
      } else {
        grid[row][col] = 0; // Backtrack
      }
    }
  }
  return false;
}

// 3. Backtracking with Forward Checking
// Forward checking is a more advanced technique.  It maintains a list of
// possible values for each empty cell.  Before assigning a value, it
// checks if that value is still possible for any unassigned cell in the
// same row, column, and block.  This can prune the search space and
// improve performance, but it adds complexity.
bool solveSudoku3(List<List<int>> grid) {
  List<List<Set<int>>> possibleValues = List.generate(
      9, (_) => List.generate(9, (_) => {1, 2, 3, 4, 5, 6, 7, 8, 9}));

  // Initialize possible values
  for (int row = 0; row < 9; row++) {
    for (int col = 0; col < 9; col++) {
      if (grid[row][col] != 0) {
        possibleValues[row][col].clear(); // No possible values for filled cells
        updatePossibleValues(grid, possibleValues, row, col,
            grid[row][col]);
      }
    }
  }

  return solveSudokuWithForwardChecking(grid, possibleValues);
}

bool solveSudokuWithForwardChecking(
    List<List<int>> grid, List<List<Set<int>>> possibleValues) {
  int row = -1;
  int col = -1;
  bool isEmpty = true;

  for (int i = 0; i < 9; i++) {
    for (int j = 0; j < 9; j++) {
      if (grid[i][j] == 0) {
        row = i;
        col = j;
        isEmpty = false;
        break;
      }
    }
    if (!isEmpty) break;
  }

  if (isEmpty) return true;

  for (int number in possibleValues[row][col].toList()) {
    // Iterate over a copy to avoid concurrent modification
    if (isValid(grid, row, col, number)) {
      grid[row][col] = number;
      List<List<Set<int>>> savedPossibleValues = List.generate(
          9,
              (i) => List.generate(9, (j) => Set.from(possibleValues[i][j]))); // Deep copy

      updatePossibleValues(grid, possibleValues, row, col, number);

      if (solveSudokuWithForwardChecking(grid, possibleValues)) {
        return true;
      } else {
        grid[row][col] = 0;
        possibleValues = savedPossibleValues; // Restore
      }
    }
  }
  return false;
}

void updatePossibleValues(List<List<int>> grid,
    List<List<Set<int>>> possibleValues, int row, int col, int number) {
  // Remove 'number' from possible values in the same row
  for (int c = 0; c < 9; c++) {
    possibleValues[row][c].remove(number);
  }
  // Remove 'number' from possible values in the same column
  for (int r = 0; r < 9; r++) {
    possibleValues[r][col].remove(number);
  }
  // Remove 'number' from possible values in the same block
  int startRow = row - row % 3;
  int startCol = col - col % 3;
  for (int r = startRow; r < startRow + 3; r++) {
    for (int c = startCol; c < startCol + 3; c++) {
      possibleValues[r][c].remove(number);
    }
  }
}

// 4. Backtracking with MRV (Minimum Remaining Values) Heuristic
// MRV is a heuristic that selects the cell with the fewest possible values
// to fill next.  This can significantly reduce the branching factor of the
// search tree.  This version combines MRV with backtracking.
bool solveSudoku4(List<List<int>> grid) {
  List<List<Set<int>>> possibleValues = List.generate(
      9, (_) => List.generate(9, (_) => {1, 2, 3, 4, 5, 6, 7, 8, 9}));

  // Initialize possible values
  for (int row = 0; row < 9; row++) {
    for (int col = 0; col < 9; col++) {
      if (grid[row][col] != 0) {
        possibleValues[row][col].clear(); // No possible values for filled cells
        updatePossibleValues(grid, possibleValues, row, col,
            grid[row][col]);
      }
    }
  }
  return solveSudokuWithMRV(grid, possibleValues);
}

bool solveSudokuWithMRV(
    List<List<int>> grid, List<List<Set<int>>> possibleValues) {
  int minRemaining = 10;
  int mrvRow = -1;
  int mrvCol = -1;

  for (int i = 0; i < 9; i++) {
    for (int j = 0; j < 9; j++) {
      if (grid[i][j] == 0) {
        int numRemaining = possibleValues[i][j].length;
        if (numRemaining < minRemaining) {
          minRemaining = numRemaining;
          mrvRow = i;
          mrvCol = j;
        }
      }
    }
  }

  if (mrvRow == -1 && mrvCol == -1) {
    return true; // No empty cells
  }

  for (int number in possibleValues[mrvRow][mrvCol].toList()) {
    if (isValid(grid, mrvRow, mrvCol, number)) {
      grid[mrvRow][mrvCol] = number;
      List<List<Set<int>>> savedPossibleValues = List.generate(
          9,
              (i) => List.generate(9, (j) => Set.from(possibleValues[i][j]))); // Deep copy
      updatePossibleValues(grid, possibleValues, mrvRow, mrvCol, number);
      if (solveSudokuWithMRV(grid, possibleValues)) {
        return true;
      } else {
        grid[mrvRow][mrvCol] = 0;
        possibleValues = savedPossibleValues;
      }
    }
  }
  return false;
}

// 5. Backtracking with Random Cell Selection
//  This approach selects a random empty cell at each step.  While not
//  generally the most efficient, it can sometimes perform well and
//  provides a contrast to the more deterministic approaches.
bool solveSudoku5(List<List<int>> grid) {
  List<List<int>> emptyCells = [];
  for (int r = 0; r < 9; r++) {
    for (int c = 0; c < 9; c++) {
      if (grid[r][c] == 0) {
        emptyCells.add([r, c]);
      }
    }
  }

  if (emptyCells.isEmpty) {
    return true; // No empty cells
  }

  // Randomly select an empty cell
  final random = Random();
  int randomIndex = random.nextInt(emptyCells.length);
  int row = emptyCells[randomIndex][0];
  int col = emptyCells[randomIndex][1];

  for (int number = 1; number <= 9; number++) {
    if (isValid(grid, row, col, number)) {
      grid[row][col] = number;
      if (solveSudoku5(grid)) {
        return true;
      } else {
        grid[row][col] = 0; // Backtrack
      }
    }
  }
  return false;
}

// Helper function to check if a number can be placed in a cell
bool isValid(List<List<int>> grid, int row, int col, int number) {
  // Check row
  for (int c = 0; c < 9; c++) {
    if (grid[row][c] == number) {
      return false;
    }
  }
  // Check column
  for (int r = 0; r < 9; r++) {
    if (grid[r][col] == number) {
      return false;
    }
  }
  // Check 3x3 block
  int startRow = row - row % 3;
  int startCol = col - col % 3;
  for (int r = startRow; r < startRow + 3; r++) {
    for (int c = startCol; c < startCol + 3; c++) {
      if (grid[r][c] == number) {
        return false;
      }
    }
  }
  return true;
}

File: Java/14-Backtracking/Algorithms/2-SudokuSolver/SudokuSolver.java
import java.util.ArrayList;
import java.util.List;

public class SudokuSolver {

    /**
     * Main method to test the Sudoku solvers.
     *
     * @param args command line arguments (not used)
     */
    public static void main(String[] args) {
        int[][] board = {
                {5, 3, 0, 0, 7, 0, 0, 0, 0},
                {6, 0, 0, 1, 9, 5, 0, 0, 0},
                {0, 9, 8, 0, 0, 0, 0, 6, 0},
                {8, 0, 0, 0, 6, 0, 0, 0, 3},
                {4, 0, 0, 8, 0, 3, 0, 0, 1},
                {7, 0, 0, 0, 2, 0, 0, 0, 6},
                {0, 6, 0, 0, 0, 0, 2, 8, 0},
                {0, 0, 0, 4, 1, 9, 0, 0, 5},
                {0, 0, 0, 0, 8, 0, 0, 7, 9}
        };

        System.out.println("Original Sudoku Board:");
        printBoard(board);

        System.out.println("\nSolving using Approach 1: Basic Backtracking");
        int[][] board1 = copyBoard(board); // Create a copy to avoid modifying the original
        if (solveSudoku1(board1)) {
            System.out.println("Solved Sudoku Board:");
            printBoard(board1);
        } else {
            System.out.println("No solution exists.");
        }

        System.out.println("\nSolving using Approach 2: Backtracking with Helper Functions");
        int[][] board2 = copyBoard(board);
        if (solveSudoku2(board2)) {
            System.out.println("Solved Sudoku Board:");
            printBoard(board2);
        } else {
            System.out.println("No solution exists.");
        }

        System.out.println("\nSolving using Approach 3: Optimized Backtracking");
        int[][] board3 = copyBoard(board);
        if (solveSudoku3(board3)) {
            System.out.println("Solved Sudoku Board:");
            printBoard(board3);
        } else {
            System.out.println("No solution exists.");
        }

        System.out.println("\nSolving using Approach 4: Backtracking with Precomputed Possibilities");
        int[][] board4 = copyBoard(board);
        if (solveSudoku4(board4)) {
            System.out.println("Solved Sudoku Board:");
            printBoard(board4);
        } else {
            System.out.println("No solution exists.");
        }

        System.out.println("\nSolving using Approach 5: Iterative Backtracking with Stack");
        int[][] board5 = copyBoard(board);
        if (solveSudoku5(board5)) {
            System.out.println("Solved Sudoku Board:");
            printBoard(board5);
        } else {
            System.out.println("No solution exists.");
        }
    }

    /**
     * Helper method to print the Sudoku board.
     *
     * @param board the Sudoku board to print
     */
    public static void printBoard(int[][] board) {
        for (int i = 0; i < 9; i++) {
            if (i % 3 == 0 && i != 0) {
                System.out.println("-----------");
            }
            for (int j = 0; j < 9; j++) {
                if (j % 3 == 0 && j != 0) {
                    System.out.print("|");
                }
                System.out.print(board[i][j] + " ");
            }
            System.out.println();
        }
    }

    /**
     * Helper method to copy the Sudoku board.  Used to avoid modifying the original board
     * when testing different solving approaches.
     * @param board The sudoku board to copy.
     * @return A new 2D array with the same values as the input board.
     */
    public static int[][] copyBoard(int[][] board) {
        int[][] newBoard = new int[9][9];
        for (int i = 0; i < 9; i++) {
            System.arraycopy(board[i], 0, newBoard[i], 0, 9);
        }
        return newBoard;
    }

    // Approach 1: Basic Backtracking
    // This is the most straightforward backtracking solution.  It iterates through
    // each empty cell and tries every possible number.  If a number is valid, it
    // recursively calls itself to fill the next empty cell.  If no number is valid,
    // it backtracks and tries a different number.
    /**
     * Solves the Sudoku puzzle using basic backtracking.
     *
     * @param board the Sudoku board to solve
     * @return true if the board is solved, false otherwise
     */
    public static boolean solveSudoku1(int[][] board) {
        for (int i = 0; i < 9; i++) {
            for (int j = 0; j < 9; j++) {
                if (board[i][j] == 0) {
                    for (int number = 1; number <= 9; number++) {
                        if (isValid(board, i, j, number)) {
                            board[i][j] = number;
                            if (solveSudoku1(board)) {
                                return true; // Found a solution
                            } else {
                                board[i][j] = 0; // Backtrack
                            }
                        }
                    }
                    return false; // No valid number found
                }
            }
        }
        return true; // Board is fully filled
    }

    /**
     * Checks if a number can be placed in a cell.
     *
     * @param board  the Sudoku board
     * @param row    the row of the cell
     * @param col    the column of the cell
     * @param number the number to check
     * @return true if the number is valid, false otherwise
     */
    public static boolean isValid(int[][] board, int row, int col, int number) {
        // Check row and column
        for (int i = 0; i < 9; i++) {
            if (board[row][i] == number || board[i][col] == number) {
                return false;
            }
        }
        // Check 3x3 subgrid
        int subgridRowStart = row - row % 3;
        int subgridColStart = col - col % 3;
        for (int i = subgridRowStart; i < subgridRowStart + 3; i++) {
            for (int j = subgridColStart; j < subgridColStart + 3; j++) {
                if (board[i][j] == number) {
                    return false;
                }
            }
        }
        return true;
    }

    // Approach 2: Backtracking with Helper Functions
    // This approach is similar to the first one, but it uses helper functions
    // to check the row, column, and subgrid, making the code more organized
    // and readable.  It doesn't improve the performance, but it's good
    // for code clarity.
    /**
     * Solves the Sudoku puzzle using backtracking with helper functions.
     *
     * @param board the Sudoku board to solve
     * @return true if the board is solved, false otherwise
     */
    public static boolean solveSudoku2(int[][] board) {
        for (int i = 0; i < 9; i++) {
            for (int j = 0; j < 9; j++) {
                if (board[i][j] == 0) {
                    for (int number = 1; number <= 9; number++) {
                        if (isValid2(board, i, j, number)) {
                            board[i][j] = number;
                            if (solveSudoku2(board)) {
                                return true;
                            } else {
                                board[i][j] = 0;
                            }
                        }
                    }
                    return false;
                }
            }
        }
        return true;
    }

    /**
     * Checks if a number can be placed in a cell, using helper functions.
     *
     * @param board  the Sudoku board
     * @param row    the row of the cell
     * @param col    the column of the cell
     * @param number the number to check
     * @return true if the number is valid, false otherwise
     */
    public static boolean isValid2(int[][] board, int row, int col, int number) {
        return !isInRow(board, row, number) &&
                !isInCol(board, col, number) &&
                !isInSubgrid(board, row, col, number);
    }

    /**
     * Checks if a number is already in the row.
     *
     * @param board  the Sudoku board
     * @param row    the row to check
     * @param number the number to check
     * @return true if the number is in the row, false otherwise
     */
    public static boolean isInRow(int[][] board, int row, int number) {
        for (int col = 0; col < 9; col++) {
            if (board[row][col] == number) {
                return true;
            }
        }
        return false;
    }

    /**
     * Checks if a number is already in the column.
     *
     * @param board  the Sudoku board
     * @param col    the column to check
     * @param number the number to check
     * @return true if the number is in the column, false otherwise
     */
    public static boolean isInCol(int[][] board, int col, int number) {
        for (int row = 0; row < 9; row++) {
            if (board[row][col] == number) {
                return true;
            }
        }
        return false;
    }

    /**
     * Checks if a number is already in the 3x3 subgrid.
     *
     * @param board the Sudoku board
     * @param row   the row of the cell
     * @param col   the column of the cell
     * @param number the number to check
     * @return true if the number is in the subgrid, false otherwise
     */
    public static boolean isInSubgrid(int[][] board, int row, int col, int number) {
        int subgridRowStart = row - row % 3;
        int subgridColStart = col - col % 3;
        for (int i = subgridRowStart; i < subgridRowStart + 3; i++) {
            for (int j = subgridColStart; j < subgridColStart + 3; j++) {
                if (board[i][j] == number) {
                    return true;
                }
            }
        }
        return false;
    }

    // Approach 3: Optimized Backtracking
    // This approach optimizes the backtracking by finding the next empty cell
    // that has the fewest possible valid numbers.  This reduces the number of
    // branches the algorithm has to explore, improving performance.
    /**
     * Solves the Sudoku puzzle using optimized backtracking.
     *
     * @param board the Sudoku board to solve
     * @return true if the board is solved, false otherwise
     */
    public static boolean solveSudoku3(int[][] board) {
        int[] emptyCell = findNextEmptyCell(board);
        if (emptyCell == null) {
            return true; // Board is solved
        }
        int row = emptyCell[0];
        int col = emptyCell[1];

        for (int number = 1; number <= 9; number++) {
            if (isValid(board, row, col, number)) {
                board[row][col] = number;
                if (solveSudoku3(board)) {
                    return true;
                } else {
                    board[row][col] = 0;
                }
            }
        }
        return false;
    }

    /**
     * Finds the next empty cell with the fewest possible valid numbers.
     *
     * @param board the Sudoku board
     * @return an array containing the row and column of the empty cell, or null if no empty cell is found
     */
    public static int[] findNextEmptyCell(int[][] board) {
        int minPossibleNumbers = 10; // Initialize with a value greater than 9
        int[] nextEmptyCell = null;

        for (int i = 0; i < 9; i++) {
            for (int j = 0; j < 9; j++) {
                if (board[i][j] == 0) {
                    int possibleNumbers = countPossibleNumbers(board, i, j);
                    if (possibleNumbers < minPossibleNumbers) {
                        minPossibleNumbers = possibleNumbers;
                        nextEmptyCell = new int[]{i, j};
                        if (possibleNumbers == 1) {
                            return nextEmptyCell; // Optimization: If only 1 possibility, return immediately
                        }
                    }
                }
            }
        }
        return nextEmptyCell;
    }

    /**
     * Counts the number of possible valid numbers for a cell.
     *
     * @param board the Sudoku board
     * @param row   the row of the cell
     * @param col   the column of the cell
     * @return the number of possible valid numbers
     */
    public static int countPossibleNumbers(int[][] board, int row, int col) {
        int count = 0;
        for (int number = 1; number <= 9; number++) {
            if (isValid(board, row, col, number)) {
                count++;
            }
        }
        return count;
    }

    // Approach 4: Backtracking with Precomputed Possibilities
    // This approach precomputes the possible numbers for each empty cell
    // before starting the backtracking.  This can speed up the process
    // because the isValid function doesn't need to be called repeatedly.
    /**
     * Solves the Sudoku puzzle using backtracking with precomputed possibilities.
     *
     * @param board the Sudoku board to solve
     * @return true if the board is solved, false otherwise
     */
    public static boolean solveSudoku4(int[][] board) {
        List<Cell> emptyCells = new ArrayList<>();
        List<List<Integer>> possibleNumbers = new ArrayList<>();

        // Precompute empty cells and their possible numbers
        for (int i = 0; i < 9; i++) {
            for (int j = 0; j < 9; j++) {
                if (board[i][j] == 0) {
                    emptyCells.add(new Cell(i, j));
                    List<Integer> numbers = new ArrayList<>();
                    for (int number = 1; number <= 9; number++) {
                        if (isValid(board, i, j, number)) {
                            numbers.add(number);
                        }
                    }
                    possibleNumbers.add(numbers);
                }
            }
        }

        return solveSudoku4Helper(board, emptyCells, possibleNumbers, 0);
    }

    /**
     * Recursive helper function for solveSudoku4.
     */
    private static boolean solveSudoku4Helper(int[][] board, List<Cell> emptyCells, List<List<Integer>> possibleNumbers, int index) {
        if (index == emptyCells.size()) {
            return true; // All cells filled
        }

        Cell cell = emptyCells.get(index);
        int row = cell.row;
        int col = cell.col;
        List<Integer> numbers = possibleNumbers.get(index);

        for (int number : numbers) {
            board[row][col] = number;
            if (solveSudoku4Helper(board, emptyCells, possibleNumbers, index + 1)) {
                return true;
            } else {
                board[row][col] = 0; // Backtrack
            }
        }
        return false;
    }

    /**
     * Inner class to represent a cell in the Sudoku board.
     */
    private static class Cell {
        int row;
        int col;

        public Cell(int row, int col) {
            this.row = row;
            this.col = col;
        }
    }

    // Approach 5: Iterative Backtracking with Stack
    // This approach uses a stack to keep track of the empty cells and the
    // numbers that have been tried.  This avoids recursion, which can be
    // more efficient in some cases.
    /**
     * Solves the Sudoku puzzle using iterative backtracking with a stack.
     *
     * @param board the Sudoku board to solve
     * @return true if the board is solved, false otherwise
     */
    public static boolean solveSudoku5(int[][] board) {
        Stack<CellNumber> stack = new Stack<>();
        for (int i = 0; i < 9; i++) {
            for (int j = 0; j < 9; j++) {
                if (board[i][j] == 0) {
                    stack.push(new CellNumber(i, j, 1)); // Start with 1
                }
            }
        }

        while (!stack.isEmpty()) {
            CellNumber current = stack.peek();
            int row = current.row;
            int col = current.col;
            int number = current.number;

            if (number > 9) {
                board[row][col] = 0; // Reset and backtrack
                stack.pop();
                if (!stack.isEmpty()) {
                    stack.peek().number++; // Increment number for the previous cell
                }
                continue;
            }

            if (isValid(board, row, col, number)) {
                board[row][col] = number;
                current.number = number + 1; // Increment for the next iteration
                // Find next empty cell
                boolean foundNext = false;
                for (int i = row; i < 9; i++) {
                    for (int j = (i == row) ? col + 1 : 0; j < 9; j++) { // start from col+1 if same row
                        if (board[i][j] == 0) {
                            stack.push(new CellNumber(i, j, 1));
                            foundNext = true;
                            break;
                        }
                    }
                    if (foundNext) break;
                }
                if (!foundNext && isBoardFull(board)) return true;

            } else {
                current.number++; // Try next number
            }
        }
        return false;
    }

    private static boolean isBoardFull(int[][] board){
        for(int i=0; i<9; i++){
            for(int j=0; j<9; j++){
                if(board[i][j] == 0) return false;
            }
        }
        return true;
    }
    /**
     * Inner class to represent a cell and the number being tried.
     */
    private static class CellNumber {
        int row;
        int col;
        int number;

        public CellNumber(int row, int col, int number) {
            this.row = row;
            this.col = col;
            this.number = number;
        }
    }

    //Custom Stack Implementation
    private static class Stack<T> {
        private List<T> stack;

        public Stack() {
            this.stack = new ArrayList<>();
        }

        public void push(T element) {
            stack.add(element);
        }

        public T pop() {
            if (isEmpty()) {
                return null; // Or throw an exception
            }
            return stack.remove(stack.size() - 1);
        }

        public T peek() {
            if (isEmpty()) {
                return null; // Or throw an exception
            }
            return stack.get(stack.size() - 1);
        }

        public boolean isEmpty() {
            return stack.isEmpty();
        }

        public int size(){
            return stack.size();
        }
    }
}

